### PR TITLE
Feature/rpc rewrite

### PR DESCRIPTION
**Pull request by dennisrijsdijk**

### CODE REVIEW

File: rpc/client.go [hunk 1]
```diff
@@ -26,212 +26,606 @@ type Transport interface {
...
-	Subscribe(event Event, args CmdArgs, handler Handler) error
-	Unsubscribe(event Event, args CmdArgs) error
-	Send(message Message, handler Handler) error
-	Close()
+type responseMessage struct {
+	data MessageData
+	err  error
 }
 
 func NewClient(clientID snowflake.ID, opts ...ConfigOpt) (Client, error) {
```
Comment (topi314) [inline rpc/client.go:34]:
```suggestion
func New(clientID snowflake.ID, opts ...ConfigOpt) (Client, error) {
```

File: rpc/client.go [hunk 2]
```diff
@@ -26,212 +26,606 @@ type Transport interface {
...
-		return nil, ctx.Err()
-	case <-client.readyChan:
+		return ctx.Err()
+	case <-c.readyChan:
 	}
+	return nil
+}
 
-	return client, nil
+func (c *Client) sendBytes(r io.Reader) error {
```
Comment (topi314) [inline rpc/client.go:85]: this has only one usage. why is this here?

File: rpc/client.go [hunk 3]
```diff
@@ -26,212 +26,606 @@ type Transport interface {
...
+	}
 
-	eventHandlers   map[Event]Handler
-	commandHandlers map[string]internalHandler
+	if res, err := c.send(Message{
+		Cmd:  CmdAuthorize,
+		Args: args,
+	}); err != nil {
+		return "", err
+	} else {
```
Comment (topi314) [inline rpc/client.go:119]: `else` = smelly

File: rpc/client.go [hunk 4]
```diff
@@ -26,212 +26,606 @@ type Transport interface {
...
+		return res.(CmdRsAuthorize).Code, nil
```
Comment (topi314) [inline rpc/client.go:120]:
```go
		return res.(CmdRsAuthorize).Code, nil
```
is there any chance `res` will not be of type `CmdRsAuthorize`?

File: rpc/client.go [hunk 5]
```diff
@@ -26,212 +26,606 @@ type Transport interface {
...
+	}); err != nil {
+		return nil, err
+	} else {
+		return res.(CmdRsGetGuilds).Guilds, nil
+	}
 }
 
-func (c *clientImpl) User() discord.User {
-	return c.user
+func (c *Client) GetChannel(channelID snowflake.ID) (CmdRsGetChannel, error) {
```
Comment (topi314) [inline rpc/client.go:165]: All these methods should return `(*T, error)`

File: rpc/cmd.go [hunk 6]
```diff
@@ -223,16 +235,21 @@ type ShortcutKeyCombo struct {
...
+	NoiseSuppression     *bool              `json:"noise_suppression,omitempty"`
+	QOS                  *bool              `json:"qos,omitempty"`
+	SilenceWarning       *bool              `json:"silence_warning,omitempty"`
+	Deaf                 *bool              `json:"deaf,omitempty"`
+	Mute                 *bool              `json:"mute,omitempty"`
+}
+
+func (EmptyArgs) cmdArgs() {}
+
+type EmptyArgs struct {
```
Comment (topi314) [inline rpc/cmd.go:252]: can't you use nil instead?

File: rpc/cmd.go [hunk 7]
```diff
@@ -131,14 +142,15 @@ func (CmdRsSetUserVoiceSettings) messageData() {}
...
+	Timeout   int          `json:"timeout,omitempty"`
 	Force     bool         `json:"force"`
+	Navigate  bool         `json:"navigate"`
 }
 
 func (CmdArgsSelectVoiceChannel) cmdArgs() {}
 
 type CmdRsSelectVoiceChannel struct {
-	PartialChannel
+	*PartialChannel
```
Comment (topi314) [inline rpc/cmd.go:153]: is this a pointer because the result can be null?

File: rpc/embed.go [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+	Thumbnail      *EmbedResource         `json:"thumbnail,omitempty"`
+	Video          *EmbedResource         `json:"video,omitempty"`
+	Provider       *discord.EmbedProvider `json:"provider,omitempty"`
+	Author         *discord.EmbedAuthor   `json:"author,omitempty"`
+}
+
+type Attachment struct {
+	*discord.Attachment
+	Spoiler bool `json:"spoiler,omitempty"`
+}
```
Comment (topi314) [inline rpc/embed.go:37]: is embed so different from `discord.Embed`?

File: rpc/message.go [hunk 9]
```diff
@@ -268,29 +269,147 @@ type CmdArgs interface {
 	cmdArgs()
 }
 
-type Event string
+type Event struct {
+	EventType
+	CmdArgs
+}
+
+func BuildIncomingEvent(eventType EventType, data MessageData) (Event, error) {
```
Comment (topi314) [inline rpc/message.go:277]:
```suggestion
func UnmarshalEvent(eventType EventType, data MessageData) (Event, error) {
```

**Review by topi314 — CHANGES_REQUESTED** [review]
in general this looks nice

I'm unsure if I like how subscribing to events work.
Maybe it should work more like in the normal bot where you register handlers via `AddEventListeners` and the `subscribe` method won't take a handler

I'm open for discussion here

File: rpc/client.go (hunk 2 above)
Comment (dennisrijsdijk) [reply rpc/client.go]: in old rpc branch this was the same (`(c *clientImpl) send(r io.Reader) error`) with only one usage. This has been inlined in [f531263](https://github.com/dennisrijsdijk/disgo/commit/f531263bd0a55277578b3e5055aef04f5bcab81d)

File: rpc/client.go (hunk 4 above)
Comment (dennisrijsdijk) [reply rpc/client.go:99]: unless there's an error (checked before accessing `res`) it's always `CmdRsAuthorize`

File: rpc/cmd.go (hunk 7 above)
Comment (dennisrijsdijk) [reply rpc/cmd.go:153]: Correct. When leaving a voice channel, the result is null.

Comment (topi314) [reply rpc/cmd.go:153]: then you wont need this to be a pointer
```suggestion
	PartialChannel
```

File: rpc/client.go (hunk 1 above)
Comment (dennisrijsdijk) [reply rpc/client.go]: fixed in [4761252](https://github.com/dennisrijsdijk/disgo/commit/47612524627647d77f406eb356b1f85dfa952f1e)

File: rpc/message.go (hunk 9 above)
Comment (dennisrijsdijk) [reply rpc/message.go]: fixed in [4761252](https://github.com/dennisrijsdijk/disgo/commit/47612524627647d77f406eb356b1f85dfa952f1e)

File: rpc/cmd.go (hunk 7 above)
Comment (dennisrijsdijk) [reply rpc/cmd.go:153]: for `CmdRsSelectVoiceChannel` that makes sense. What about `CmdRsGetSelectedVoiceChannel`? Should that remain a pointer (`if channel == nil` to detect no channel) or should I also remove the pointer there?  (`if channel.ID == 0` to detect no channel)

Comment (topi314) [reply rpc/cmd.go:153]: that shouldnt be a pointer either
the struct will be nil in this case

Comment (dennisrijsdijk) [reply rpc/cmd.go:153]: structs can't be `nil` though. it defaults to a struct with all 0 values

Comment (topi314) [reply rpc/cmd.go:153]: ah right we don't use a pointer here https://github.com/dennisrijsdijk/disgo/blob/feature/rpc-rewrite/rpc/message.go#L188
then this is fine and we can check for `id == 0`

Comment (dennisrijsdijk) [reply rpc/cmd.go:153]: Then again, I'm working on the earlier comment that everything should return `(*T, error)` so I could add that logic into the function there as well.

Comment (topi314) [reply rpc/cmd.go:153]: sounds good to me

File: rpc/client.go (hunk 5 above)
Comment (dennisrijsdijk) [reply rpc/client.go:144]: Did you want this on Authorize as well?
eg. from `Authorize(scopes []discord.OAuth2Scope, rpcToken string, username string) (string, error)`
to `Authorize(scopes []discord.OAuth2Scope, rpcToken string, username string) (*string, error)`
or `Authorize(scopes []discord.OAuth2Scope, rpcToken string, username string) (*CmdRsAuthorize, error)`

Comment (topi314) [reply rpc/client.go:144]: yes

Comment (dennisrijsdijk) [reply rpc/client.go]: Resolved in [4ef8ffc](https://github.com/dennisrijsdijk/disgo/commit/4ef8ffc03ba58f70cb22951e155ddf66e18ef4e6)

File: rpc/cmd.go (hunk 7 above)
Comment (dennisrijsdijk) [reply rpc/cmd.go]: Resolved in [4ef8ffc](https://github.com/dennisrijsdijk/disgo/commit/4ef8ffc03ba58f70cb22951e155ddf66e18ef4e6)

File: rpc/client.go (hunk 3 above)
Comment (dennisrijsdijk) [reply rpc/client.go]: Resolved in [4ef8ffc](https://github.com/dennisrijsdijk/disgo/commit/4ef8ffc03ba58f70cb22951e155ddf66e18ef4e6)

File: rpc/cmd.go (hunk 6 above)
Comment (dennisrijsdijk) [reply rpc/cmd.go]: Resolved in [ca68a83](https://github.com/dennisrijsdijk/disgo/commit/ca68a83591c1ade9626007093a8638ccebbde717)

File: rpc/embed.go (hunk 8 above)
Comment (dennisrijsdijk) [reply rpc/embed.go]: changed in [fb13deb](https://github.com/dennisrijsdijk/disgo/commit/fb13deb455f73a7061e86ba02f7294bd2bd87020), looking for more feedback.

File: rpc/message.go [hunk 10]
```diff
@@ -268,29 +339,147 @@ type CmdArgs interface {
 	cmdArgs()
 }
 
-type Event string
+type Event struct {
```
Comment (topi314) [inline rpc/message.go:342]: I'm not sure what the purpose of this is

**Review by topi314 — CHANGES_REQUESTED** [review]
I think we need to talk about the event subscriptions

File: rpc/message.go [hunk 11]
```diff
@@ -1,26 +1,97 @@
...
+	m.Attachments = v.Attachments
+	m.Blocked = v.Blocked
+	m.Bot = v.Bot
+	m.Embeds = v.Embeds
+	m.Mentions = v.Mentions
+	m.Nick = v.Nick
+	return nil
+}
+
+type message struct {
```
[the remark this answers is not in the document]
Comment (dennisrijsdijk) [reply rpc/message.go:80]: Would you suggest I rename this or Message (as in ChannelMessage)?

File: rpc/message.go (hunk 10 above)
Comment (dennisrijsdijk) [reply rpc/message.go:342]: Related atm to keeping track of subscribed events (event name + args) and handlers. to be revisited when we decide on the approach for events going forward.
